File: libs/java/src/main/java/com/mbreissi/edgecommons/metrics/targets/Messaging.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.metrics.targets;

import com.mbreissi.edgecommons.config.ConfigManager;
import com.mbreissi.edgecommons.messaging.Message;
import com.mbreissi.edgecommons.messaging.MessageBuilder;
import com.mbreissi.edgecommons.messaging.MessageIdentity;
import com.mbreissi.edgecommons.messaging.MessagingClient;
import com.mbreissi.edgecommons.messaging.Qos;
import com.mbreissi.edgecommons.metrics.Metric;
import com.mbreissi.edgecommons.uns.Uns;
import com.mbreissi.edgecommons.uns.UnsClass;
import com.google.gson.JsonObject;

import java.util.Map;

/**
 * The {@code messaging} metric target (UNS-CANONICAL-DESIGN §4.3): publishes each metric to the
 * library-owned UNS metric topic {@code ecv1/{device}/{component}/main/metric/{metricName}} (the
 * metric name sanitized as a channel token) through the privileged
 * {@link com.mbreissi.edgecommons.messaging.ReservedPublisher} seam — the {@code metric} class is
 * reserved. {@code metricEmission.targetConfig.destination} still selects local/IPC vs northbound
 * (D-U9); the legacy {@code targetConfig.topic} override is removed.
 */
public final class Messaging extends MetricTarget {

    private boolean sendToIpc = true;
    private MessagingClient messagingService;
    /** WARN-once flag for the no-resolved-identity (test/subclass bring-up) case. */
    private boolean warnedNoIdentity = false;

    public Messaging(ConfigManager configManager) {
        super(configManager);
        this.sendToIpc = !isNorthboundDestination(metricConfig.getDestination());
    }

    /**
     * Northbound is selected only by "northbound"; everything else (the canonical "ipc",
     * "local", and any unrecognized value) uses the local/IPC transport, so a metric never routes
     * to a possibly-unconfigured northbound broker. Matches the heartbeat destination and the
     * Python/Rust metric targets.
     */
    private static boolean isNorthboundDestination(String destination) {
        return destination != null && destination.equalsIgnoreCase("northbound");
    }

    public void setMessagingService(MessagingClient messagingService) {
        this.messagingService = messagingService;
    }

    @Override
    public void emitMetricNow(Metric metric, Map<String, Float> measureValues)
    {
        JsonObject metricObject = EmfHelper.buildMetricData(metricConfig.getNamespace(), metric, measureValues, false);
        publishMessage(metric, metricObject);

        if (metricConfig.getLargeFleetWorkaround())
        {
            metricObject = EmfHelper.buildMetricData(metricConfig.getNamespace(), metric, measureValues, true);
            publishMessage(metric, metricObject);
        }
    }

    /**
     * The metric's UNS topic — {@code ecv1[/{site}]/{device}/{component}/main/metric/{name}} with
     * the metric name passed through the template sanitizer (the §2.2 channel-token rule), or
     * {@code null} (WARN once) when no component identity is resolved (mock/test bring-up).
     */
    private String metricTopic(Metric metric)
    {
        MessageIdentity identity = configManager.getComponentIdentity();
        if (identity == null)
        {
            if (!warnedNoIdentity)
            {
                warnedNoIdentity = true;
                LOGGER.warn("No resolved component identity - the messaging metric target cannot"
                        + " build UNS metric topics; metrics are dropped");
            }
            return null;
        }
        return new Uns(identity, configManager.isTopicIncludeRoot())
                .topic(UnsClass.METRIC, ConfigManager.sanitize(metric.getName()));
    }

    private void publishMessage(Metric metric, JsonObject metricObject)
    {
        String topic = metricTopic(metric);
        if (topic == null)
        {
            return;
        }
        Message message = MessageBuilder.create("Metric", "1.0")
                .withPayload(metricObject)
                .withConfig(configManager)
                .build();
        // The metric class is reserved (§4.1) - publish through the privileged seam (§4.2).
        if (sendToIpc)
            messagingService.reservedPublisher().publish(topic, message);
        else
            messagingService.reservedPublisher().publishNorthbound(topic, message, Qos.AT_LEAST_ONCE);
        LOGGER.trace("Metric emitted for {} emitted", metric);
    }

    @Override
    public boolean onConfigurationChanged()
    {
        LOGGER.info("Configuration changed. Reconfiguring metric messaging destination");
        this.sendToIpc = !isNorthboundDestination(configManager.getMetricConfig().getDestination());
        return true;
    }

    @Override
    public void emitMetric(Metric metric, Map<String, Float> measureValues)
    {
        emitMetricNow(metric, measureValues);
    }
}

File: libs/python/edgecommons/metrics/targets/messaging.py
from edgecommons.messaging.message_builder import MessageBuilder
from edgecommons.messaging.messaging_client import MessagingClient
from edgecommons.messaging.qos import Qos
from edgecommons.config.manager.config_manager import ConfigManager
from edgecommons.metrics.targets.emf_helper import build_metric_data_emf
from edgecommons.metrics.targets.metric_target import MetricTarget
from edgecommons.uns import Uns, UnsClass


def _is_local_destination(destination: str) -> bool:
    """True for the local/IPC transport, False for northbound.

    Northbound is selected only by "northbound"; everything else ("ipc", "local", and any
    unrecognized value) uses the local transport, so a metric never fails by routing to an
    unconfigured northbound broker. Matches the Java/Rust metric targets and the config schema.
    """
    return destination is None or destination.lower() != "northbound"


class Messaging(MetricTarget):
    """The ``messaging`` metric target (UNS-CANONICAL-DESIGN §4.3): publishes each
    metric to the library-owned UNS metric topic
    ``ecv1/{device}/{component}/main/metric/{metricName}`` (the metric name sanitized
    as a channel token) through the privileged ``MessagingClient._publish_reserved*``
    seam — the ``metric`` class is reserved. ``metricEmission.targetConfig.destination``
    still selects local/IPC vs northbound (D-U9); the legacy ``targetConfig.topic``
    override is removed."""

    def __init__(self, config_manager: ConfigManager):
        super().__init__(config_manager)
        self.send_to_local = _is_local_destination(
            config_manager.get_metric_config().get_destination()
        )
        # WARN-once flag for the no-resolved-identity (test/subclass bring-up) case.
        self._warned_no_identity = False

    def emit_metric_now(self, metric, measure_values):
        metric_name = metric.get_name()
        self.logger.debug(f"Emitting metric '{metric_name}' to messaging target with {len(measure_values)} measures")

        topic = self._metric_topic(metric_name)
        if topic is None:
            return

        metric_data = build_metric_data_emf(
            self.metric_config, metric, measure_values, False
        )
        self.__publish_message(topic, metric_data)

        if self.metric_config.get_large_fleet_workaround():
            self.logger.debug(f"Emitting large fleet workaround metric for '{metric_name}'")
            metric_data = build_metric_data_emf(
                self.metric_config, metric, measure_values, True
            )
            self.__publish_message(topic, metric_data)

        self.logger.debug(f"Metric '{metric_name}' emission completed")

    def _metric_topic(self, metric_name: str):
        """The metric's UNS topic —
        ``ecv1[/{site}]/{device}/{component}/metric/{name}`` (component scope, no
        instance — D-U28) with the metric name passed through the template sanitizer
        (the §2.2 channel-token rule), or
        ``None`` (WARN once) when no component identity is resolved (mock/test
        bring-up)."""
        identity = self.config_manager.get_component_identity()
        if identity is None:
            if not self._warned_no_identity:
                self._warned_no_identity = True
                self.logger.warning(
                    "No resolved component identity - the messaging metric target"
                    " cannot build UNS metric topics; metrics are dropped"
                )
            return None
        return Uns(identity, self.config_manager.is_topic_include_root()).topic(
            UnsClass.METRIC, ConfigManager.sanitize(metric_name)
        )

    def __publish_message(self, topic: str, metric_dict: dict):
        destination = "local" if self.send_to_local else "northbound"
        self.logger.debug(f"Publishing metric message to {destination} on topic: {topic}")

        message = MessageBuilder.create("Metric", "1.0") \
            .with_payload(metric_dict) \
            .with_config(self.config_manager) \
            .build()

        # The metric class is reserved (§4.1) - publish through the privileged seam
        # (§4.2).
        if self.send_to_local:
            MessagingClient._publish_reserved(topic, message)
        else:
            MessagingClient._publish_reserved_northbound(topic, message, Qos.AT_LEAST_ONCE)

    def on_configuration_change(self, configuration) -> bool:
        self.logger.info("Metric messaging configuration changed, reconfiguring target")

        old_destination = "local" if self.send_to_local else "northbound"
        self.send_to_local = _is_local_destination(
            self.config_manager.get_metric_config().get_destination()
        )
        new_destination = "local" if self.send_to_local else "northbound"

        self.logger.info(f"Metric messaging reconfigured - destination: {old_destination} -> {new_destination}")
        return True
